src/disco/shred/fd_fec_resolver.c

```diff
@@ -0,0 +1,527 @@
+  if( FD_UNLIKELY( !freelist_new ( cmplst, complete_depth+1UL       )) ) { FD_LOG_WARNING(("freelist_new failed")); return NULL; }
+  if( FD_UNLIKELY( !bmtrlist_new ( bmfree, depth                    )) ) { FD_LOG_WARNING(("bmtrlist_new failed")); return NULL; }
+  if( FD_UNLIKELY( !fd_sha512_new( (void *)resolver->sha512         )) ) { FD_LOG_WARNING(("sha512_new failed"  )); return NULL; }
+
+  /* Initialize all the lists */
+  fd_fec_set_t * * free_list     = freelist_join( free   );
+  fd_fec_set_t * * complete_list = freelist_join( cmplst );
+  for( ulong i=0UL;                 i<depth+partial_depth;                i++ ) { freelist_push_tail( free_list,     sets+i ); }
+  for( ulong i=depth+partial_depth; i<depth+partial_depth+complete_depth; i++ ) { freelist_push_tail( complete_list, sets+i ); }
+  freelist_leave( complete_list );
+  freelist_leave( free_list     );
+
+  void * * bmtree_list = bmtrlist_join( bmfree );
+  for( ulong i=0UL; i<depth; i++ ) { bmtrlist_push_tail( bmtree_list, (uchar *)bmfootprint + i*footprint_per_bmtree ); }
+  bmtrlist_leave( bmtree_list );
+
+
+  resolver->curr_head_tag = 0UL;
+  resolver->curr_tail_tag = 0UL;
+
+  resolver->depth          = depth;
+  resolver->partial_depth  = partial_depth;
+  resolver->complete_depth = complete_depth;
+  resolver->done_depth     = done_depth;
+  return shmem;
+}
+
+fd_fec_resolver_t *
+fd_fec_resolver_join( void * shmem ) {
+  fd_fec_resolver_t * resolver = (fd_fec_resolver_t *)shmem;
+  ulong depth          = resolver->depth;
+  ulong partial_depth  = resolver->partial_depth;
+  ulong complete_depth = resolver->complete_depth;
+  ulong done_depth     = resolver->done_depth;
+
+  int lg_depth = fd_ulong_find_msb( depth );
+  ulong done_map_cnt = fd_tcache_map_cnt_default( done_depth );
+
+  FD_SCRATCH_ALLOC_INIT( l, shmem );
+  /*     self       */ FD_SCRATCH_ALLOC_APPEND( l, FD_FEC_RESOLVER_ALIGN,  sizeof(fd_fec_resolver_t)                       );
+  void * done        = FD_SCRATCH_ALLOC_APPEND( l, FD_TCACHE_ALIGN,        fd_tcache_footprint( done_depth, done_map_cnt ) );
+  void * curr        = FD_SCRATCH_ALLOC_APPEND( l, ctx_map_align(),        ctx_map_footprint( lg_depth+1 )                 );
+  void * free        = FD_SCRATCH_ALLOC_APPEND( l, freelist_align(),       freelist_footprint( depth+partial_depth+1UL )   );
+  void * cmplst      = FD_SCRATCH_ALLOC_APPEND( l, freelist_align(),       freelist_footprint( complete_depth+1UL  )       );
+  void * bmfree      = FD_SCRATCH_ALLOC_APPEND( l, bmtrlist_align(),       bmtrlist_footprint( depth+1UL )                 );
+  FD_SCRATCH_ALLOC_FINI( l, FD_FEC_RESOLVER_ALIGN );
+
+  resolver->done             = fd_tcache_join( done   );
+  resolver->curr_map         = ctx_map_join  ( curr   );
+  resolver->free_list        = freelist_join ( free   );
+  resolver->complete_list    = freelist_join ( cmplst );
+  resolver->bmtree_free_list = bmtrlist_join ( bmfree );
+  fd_sha512_join( resolver->sha512 );
```

Comment:
I think the API of join is that you will return NULL here if these inner joins fail?

Reply:
Fixed